### ISSUE TITLE
Controlling the loading order of plugins

### ISSUE THREAD

**Issue by suresttexas00** [benbaptist/minecraft-wrapper#253]
OK, Operating system change issue number 2...

  On WIndows, wrapper loaded the plugins alphabetically.  Therefore plugin "global.py" loaded prior to "stafftools.py" and "simple.py".. now it is loading AFTER them... and since they use the getplugincontext (depend on global.py), they error out during the plugin loading:
```
[18:25:10] [Wrapper.py/INFO] Loading plugins...
[18:25:10] [Wrapper.py/INFO] Loading plugin stafftools.py...
[18:25:10] [Wrapper.py/DEBUG] Traceback (most recent call last):
[18:25:10] [Wrapper.py/DEBUG]   File "/home/surest/vanilla/wrapper/plugins.py", line 93, in loadPlugins
[18:25:10] [Wrapper.py/DEBUG]     elif i[-3:] == ".py": self.loadPlugin(i)
[18:25:10] [Wrapper.py/DEBUG]   File "/home/surest/vanilla/wrapper/plugins.py", line 56, in loadPlugin
[18:25:10] [Wrapper.py/DEBUG]     main = plugin.Main(API(self.wrapper, name, id), PluginLog(self.log, name))
[18:25:10] [Wrapper.py/DEBUG]   File "wrapper-plugins/stafftools.py", line 25, in __init__
[18:25:10] [Wrapper.py/DEBUG]     self.globalset = api.getPluginContext("com.suresttexas00.plugins.global")
[18:25:10] [Wrapper.py/DEBUG]   File "/home/surest/vanilla/wrapper/api/__init__.py", line 136, in getPluginContext
[18:25:10] [Wrapper.py/DEBUG]     raise NonExistentPlugin("Plugin %s does not exist!" % id)
[18:25:10] [Wrapper.py/DEBUG] NonExistentPlugin: Plugin com.suresttexas00.plugins.global does not exist!
[18:25:10] [Wrapper.py/DEBUG] 
[18:25:10] [Wrapper.py/ERROR] Failed to import plugin 'stafftools.py'
[18:25:10] [Wrapper.py/INFO] Loading plugin voterewards.py...
[18:25:10] [Wrapper.py/INFO] Loading plugin broadcaster.py...
[18:25:10] [Wrapper.py/DEBUG] [Broadcaster] Registered event 'timer.second'
[18:25:10] [Wrapper.py/INFO] Loading plugin regions.py...
[18:25:10] [Wrapper.py/DEBUG] [Regions] Registered help group 'Regions' with 14 commands
....
[18:37:49] [Wrapper.py/INFO] Loading plugin simple.py...
[18:37:49] [Wrapper.py/DEBUG] Traceback (most recent call last):
[18:37:49] [Wrapper.py/DEBUG]   File "/home/surest/vanilla/wrapper/plugins.py", line 93, in loadPlugins
[18:37:49] [Wrapper.py/DEBUG]     elif i[-3:] == ".py": self.loadPlugin(i)
[18:37:49] [Wrapper.py/DEBUG]   File "/home/surest/vanilla/wrapper/plugins.py", line 56, in loadPlugin
[18:37:49] [Wrapper.py/DEBUG]     main = plugin.Main(API(self.wrapper, name, id), PluginLog(self.log, name))
[18:37:49] [Wrapper.py/DEBUG]   File "wrapper-plugins/simple.py", line 23, in __init__
[18:37:49] [Wrapper.py/DEBUG]     self.globalset = api.getPluginContext("com.suresttexas00.plugins.global")
[18:37:49] [Wrapper.py/DEBUG]   File "/home/surest/vanilla/wrapper/api/__init__.py", line 136, in getPluginContext
[18:37:49] [Wrapper.py/DEBUG]     raise NonExistentPlugin("Plugin %s does not exist!" % id)
[18:37:49] [Wrapper.py/DEBUG] NonExistentPlugin: Plugin com.suresttexas00.plugins.global does not exist!
[18:37:49] [Wrapper.py/DEBUG] 
[18:37:49] [Wrapper.py/ERROR] Failed to import plugin 'simple.py'
[18:37:49] [Wrapper.py/INFO] Loading plugin global.py...
Global [[18:37:49]/DEBUG] passed up on bad item: MaxPlayersPerIP = 3
Global [[18:37:49]/DEBUG] NewPlayerMessage = &aWelcome to the server %s.  Type /help to begin!
Global [[18:37:49]/DEBUG] SlashSpawnCoords = 3259.5,104,2669.5,0,0
Global [[18:37:49]/DEBUG] toolbox_useVanillaBehavior = false
Global [[18:37:49]/DEBUG] useGlobalChat = true
Global [[18:37:49]/DEBUG] teleport_useVanillaBehavior = true
Global [[18:37:49]/DEBUG] ReturningPlayerMessage = &aWelcome back, %s! Most /help commands are clickable.
Global [[18:37:49]/DEBUG] toolbox_usePermissions = true
Global [[18:37:49]/DEBUG] claims_useVanillaBehavior = false
Global [[18:37:49]/DEBUG] teleport_usePermissions = true
Global [[18:37:49]/DEBUG] defaultOp = SurestTexas00
Global [[18:37:49]/DEBUG] /n
[18:37:49] [Wrapper.py/DEBUG] [Global] Registered command 'global'
[18:37:49] [Wrapper.py/DEBUG] [Global] Registered event 'timer.second'
[18:37:49] [Wrapper.py/DEBUG] [Global] Registered event 'player.login'
[18:37:49] [Wrapper.py/DEBUG] [Global] Registered event 'player.death'
[18:37:49] [Wrapper.py/DEBUG] [Global] Registered event 'player.rawMessage'
[18:37:49] [Wrapper.py/DEBUG] [Global] Registered event 'player.chatbox'
[18:37:49] [Wrapper.py/INFO] Loading plugin portals.py...
...
```

**Comment by suresttexas00** [comment]
The plugins are listed below.  They load in this order:
stafftools 
voterewards
broadcaster
regions
surestlib
teleports
vanillaclaims
simple
global
portals
toolbox
voting
votiPyer
signs

![image](https://cloud.githubusercontent.com/assets/10001762/12076641/bdaf2a58-b181-11e5-8af3-245a271bec44.png)

**Comment by JasonBristol** [comment]
This has my vote! I have encountered this as well on my Linux server. I suspected that it was the load order, but couldn't be completely sure without a successful use case.

**Comment by suresttexas00** [comment]
I fixed this with pull request #254

**Comment by suresttexas00** [comment]
Changed to (slightly better) pull request #257

**Comment by benbaptist** [comment]
Good point. Does adding a number in the beginning of the file force a different loading order?
```
0_sandwich.py
1_pizza.py
2_whatever_else.py
```
Although, it doesn't look like it's really loading them alphabetically anyways, but maybe it'll influence it. I think it's just going by however os.listdir lists them.

**Comment by benbaptist** [comment]
I'll submit a patch to use sort() to ensure a more logical loading order. This means it'll load alphabetically - prefix your plugin files with a number to force ordering. (similar to those weird config files in Xorg, or ini init.d, etc.)

**Comment by suresttexas00** [comment]
Why not use the dependency loader?  I don't particularly want to rename plugins just so that they load in order.

**Comment by benbaptist** [comment]
You aren't actually naming it themselves - just prefixing the filename with ah integer.
```
0_global.py
10_olives.py
20_meowmix.py
```

But a dependency-based loader would be far more superior. I looked at the pull request, and I'll accept it but maybe we should rewrite ALL the things. It made me barf looking at how I handled getting variables - `try: plugin.BLAH` !?!? What on earth was I on? It's overall just ugly code, so I'd like to refactor it to just make it a bit nicer.

**Comment by suresttexas00** [comment]
Thanks.  I know the "internal" plugin name is not changed; I just don't like the hacky look of making filenames different from the plugin and with all kinds of number prefixes..   If you did use a sort(), my plugins _would_ work without the numbers _as-is_ because that is how they were designed when I was on Wind:sleeping:ze ... However, I do rather like the freedom to choose filenames irrespective of their sorting order.

**Comment by Sasszem** [comment]
We should have a tool to set the order of processing events, because if we've got 10 plugins that handle `slotClick` events, if the first returns that we can forward it to the serv, then the second don't get it, so it's missed.

**Comment by suresttexas00** [comment]
I believe (based on experience) that the event is passed to all the plugins with that event.  If plugin "A" says it can be returned, but plugin "B" says that it cannot, the the event gets canceled.. at least that is the behavior I have observed with player.Place and player.Dig Events...  We should really test this behavior.  I generally avoid having plugins that do the same thing.  So plugin A will handle chat type events , plugin C will handle all block digging and placing, etc...

**Comment by suresttexas00** [comment]
FYI, Sasszem, event processing is another issue... maybe open a different issue/suggestion (after verifying whether or not a second plugin "misses" events?

**Comment by suresttexas00** [comment]
Don't bother with PR #257 as it is included with #275.

**Comment by suresttexas00** [comment]
re-opening the discussion:  I tested Sasszem's proposition that event ordering may be an issue.  I have not until now made or tested plugins where an event like player.rawMessage would be handled in more than one plugin.  

After some testing tonight, I discovered that the `player.rawMessage` event will _only_ go to one of the registered events (presumably the one that registered first?).  And, yes, I tested this against wrapper 7.7 build 110.
